### PR TITLE
Migrate spacy featurizer for the graph.

### LINKED ISSUE #9452: Migrate `SpacyFeaturizer ` to GraphComponent interface
**Overview of the Solution**:
Adapt the `SpacyFeaturizer ` to the new `GraphComponent ` interface.
See the [implementation proposal](https://www.notion.so/rasa/Rasa-Open-Source-3-0-Architecture-Implementation-Proposal-51ab90b05c41435ca98189a101676a1e) for further details.

* adapt `create` and `load` functions to new signatures
* implement `required_packages` and `supported_languages` 
* rename `train` to `process_training_data` (this featurizer only adds features but doesn't require training)
* support multiple messages in `process` (**make sure to handle empty messages**)

**Definition of Done**:
- [ ] SpacyFeaturizer adapted to the new GraphComponent interface
- [ ] Unit tests are adapted to adapted component
- [ ] Other tests (integration tests, etc.) don't break (because they'd (for now) still use the old component implementation)

### PR DESCRIPTION
**Pull request by joejuzl**
Closes #9452.

### CODE REVIEW

File: rasa/nlu/featurizers/dense_featurizer/spacy_featurizer.py [hunk 1]
```diff
@@ -18,55 +17,64 @@
...
+            # Specify what pooling operation should be used to calculate the vector of
+            # the complete utterance. Available options: 'mean' and 'max'
+            POOLING: MEAN_POOLING,
+        }
 
-    def __init__(self, component_config: Optional[Dict[Text, Any]] = None):
-        super().__init__(component_config)
+    def __init__(
+        self, config: Dict[Text, Any], name: Text,
+    ):
```
Comment (wochinge) [inline rasa/nlu/featurizers/dense_featurizer/spacy_featurizer.py:45]:
```suggestion
    ) -> None:
```

File: rasa/nlu/featurizers/dense_featurizer/spacy_featurizer.py [hunk 2]
```diff
@@ -77,21 +85,33 @@ def _set_spacy_features(self, message: Message, attribute: Text = TEXT) -> None:
...
+    def validate_config(cls, config: Dict[Text, Any]) -> None:
+        """Validates that the component is configured properly."""
+        pass
+
+    @classmethod
+    def validate_compatibility_with_tokenizer(
+        cls, config: Dict[Text, Any], tokenizer_type: Type[Tokenizer]
+    ) -> None:
+        """Validates that the featurizer is compatible with the given tokenizer."""
+        pass
```
Comment (wochinge) [inline rasa/nlu/featurizers/dense_featurizer/spacy_featurizer.py:117]: can we actually check that it's the `SpacyTokenizer`?

File: tests/nlu/featurizers/test_spacy_featurizer.py [hunk 3]
```diff
@@ -1,23 +1,33 @@
...
-def spacy_nlp_component(component_builder, blank_config):
-    spacy_nlp_config = {"name": "SpacyNLP", "model": "en_core_web_md"}
-    return component_builder.create_component(spacy_nlp_config, blank_config)
+def spacy_model() -> SpacyModel:
+    return SpacyModelProvider.create(
+        {"model": "en_core_web_md"}, None, None, None
+    ).provide()
+
+
+def create_spacy_featurizer(config):
```
Comment (wochinge) [inline tests/nlu/featurizers/test_spacy_featurizer.py:22]:
```suggestion
def create_spacy_featurizer(config: Dict[Text, Any) -> SpacyFeaturizerGraphComponent:
```

File: tests/nlu/featurizers/test_spacy_featurizer.py [hunk 4]
```diff
@@ -52,15 +62,17 @@ def test_spacy_featurizer(sentence, spacy_nlp):
     assert np.allclose(vecs, expected, atol=1e-5)
 
 
-def test_spacy_training_sample_alignment(spacy_nlp_component):
+def test_spacy_training_sample_alignment(spacy_model):
```
Comment (wochinge) [inline tests/nlu/featurizers/test_spacy_featurizer.py:65]:
```suggestion
def test_spacy_training_sample_alignment(spacy_model: SpacyModel):
```

File: tests/nlu/featurizers/test_spacy_featurizer.py [hunk 5]
```diff
@@ -1,23 +1,33 @@
...
 from rasa.shared.nlu.constants import TEXT, INTENT, RESPONSE
 
 
 @pytest.fixture(scope="module")
-def spacy_nlp_component(component_builder, blank_config):
-    spacy_nlp_config = {"name": "SpacyNLP", "model": "en_core_web_md"}
-    return component_builder.create_component(spacy_nlp_config, blank_config)
+def spacy_model() -> SpacyModel:
+    return SpacyModelProvider.create(
+        {"model": "en_core_web_md"}, None, None, None
```
Comment (wochinge) [inline tests/nlu/featurizers/test_spacy_featurizer.py:18]: nit: seems a bit evil to pass `None` if `None` isn't officially allowed, what do you think?

File: rasa/nlu/featurizers/dense_featurizer/spacy_featurizer.py [hunk 6]
```diff
@@ -18,55 +17,64 @@
...
-                self._set_spacy_features(example, attribute)
-
-    def get_doc(self, message: Message, attribute: Text) -> Any:
+    def _get_doc(self, message: Message, attribute: Text) -> Any:
         return message.get(SPACY_DOCS[attribute])
 
-    def process(self, message: Message, **kwargs: Any) -> None:
-        for attribute in DENSE_FEATURIZABLE_ATTRIBUTES:
-            self._set_spacy_features(message, attribute)
+    def process(self, messages: List[Message], **kwargs: Any) -> List[Message]:
```
Comment (wochinge) [inline rasa/nlu/featurizers/dense_featurizer/spacy_featurizer.py:68]:
```suggestion
    def process(self, messages: List[Message]) -> List[Message]:
```

File: rasa/nlu/featurizers/dense_featurizer/spacy_featurizer.py [hunk 7]
```diff
@@ -18,55 +17,64 @@
...
+        """Process incoming messages and compute and set features."""
```
Comment (wochinge) [inline rasa/nlu/featurizers/dense_featurizer/spacy_featurizer.py:69]:
```suggestion
        """Processes incoming messages and computes and sets features."""
```

File: rasa/nlu/featurizers/dense_featurizer/spacy_featurizer.py (hunk 1 above)
Comment (joejuzl) [reply rasa/nlu/featurizers/dense_featurizer/spacy_featurizer.py:45]: every time 🤦

File: tests/nlu/featurizers/test_spacy_featurizer.py (hunk 5 above)
Comment (joejuzl) [reply tests/nlu/featurizers/test_spacy_featurizer.py:18]: yeah true, I can just use the `__init__` instead.

Comment (joejuzl) [reply tests/nlu/featurizers/test_spacy_featurizer.py:18]: yeah true, it's just annoying have to pass the fixtures around everywhere.

File: rasa/nlu/featurizers/dense_featurizer/spacy_featurizer.py [hunk 8]
```diff
@@ -77,21 +85,36 @@ def _set_spacy_features(self, message: Message, attribute: Text = TEXT) -> None:
...
+        pass
+
+    @classmethod
+    def validate_compatibility_with_tokenizer(
+        cls, config: Dict[Text, Any], tokenizer_type: Type[Tokenizer]
+    ) -> None:
+        """Validates that the featurizer is compatible with the given tokenizer."""
+        if tokenizer_type != SpacyTokenizer:
+            rasa.shared.utils.io.raise_warning(
+                f"Expected {SpacyTokenizer.__name__} to be present."
```
Comment (wochinge) [inline rasa/nlu/featurizers/dense_featurizer/spacy_featurizer.py:119]:
```suggestion
                f"Expected '{SpacyTokenizer.__name__}' to be present."
```

File: tests/nlu/featurizers/test_spacy_featurizer.py [hunk 9]
```diff
@@ -1,23 +1,41 @@
...
+) -> SpacyModel:
+    return SpacyModelProvider.create(
+        {"model": "en_core_web_md"},
+        default_model_storage,
+        Resource("spacy_model_provider"),
+        default_execution_context,
+    ).provide()
+
+
+def create_spacy_featurizer(config) -> SpacyFeaturizerGraphComponent:
```
Comment (wochinge) [inline tests/nlu/featurizers/test_spacy_featurizer.py:30]:
```suggestion
def create_spacy_featurizer(config: Dict[Text, Any]) -> SpacyFeaturizerGraphComponent:
```

File: rasa/nlu/featurizers/dense_featurizer/spacy_featurizer.py [hunk 10]
```diff
@@ -77,21 +85,36 @@ def _set_spacy_features(self, message: Message, attribute: Text = TEXT) -> None:
...
         )
         message.add_features(final_sentence_features)
+
+    @classmethod
+    def validate_config(cls, config: Dict[Text, Any]) -> None:
+        """Validates that the component is configured properly."""
+        pass
+
+    @classmethod
+    def validate_compatibility_with_tokenizer(
```
Comment (wochinge) [inline rasa/nlu/featurizers/dense_featurizer/spacy_featurizer.py:113]: can we please add a test for this?

**Review by wochinge — COMMENTED** [review]
🚀

**Review by wochinge — APPROVED** [review]
Sorry, hit the wrong button 😀

**Pull request merged**